### PR TITLE
fix: timestamps mismatch

### LINKED ISSUE #468: fix: transaction timestamp not matching block timestamp
# Example

- `Jul 17, 2025 at 06:32:16 GMT+2` block timestamp [here](http://localhost:3000/block/956255?cluster=custom&customUrl=https%3A%2F%2Fdevnet.magicblock.app)
- `Jul 17, 2025 at 06:42:48 Central European Summer Time` transaction timestamp [here](http://localhost:3000/tx/3FSXi1Ho6tA8WmbrNAreD7YmmYnj5muyRS2xJxnEY3eDaZjBuj2KVhxQNjbmArFMU8qNeHUGqWAxgepBbjdZ3V4M?cluster=custom&customUrl=https%3A%2F%2Fdevnet.magicblock.app)

### LINKED ISSUE #469: fix: ledger clock does not match the clock used by the validator
Some users are facing cases where replaying transactions yields different results than what the validator got. This is caused by the clock sysvar having different values for the replay and in the ledger

### PR DESCRIPTION
**Pull request by Dodecahedr0x**
Closes #468 and closes #469

Explorer was showing the transaction timestamp because it uses `getBlockTime`, which was returning an estimate of the slot time based on the slot period and the number of slots between the current slot and the requested slot. This is fixed by returning the timestamp of the block from the ledger.

The timestamp written in the ledger was a different system time from the one used for the clock. This is fixed by writing in the ledger the latest value of the clock, which should also fix replay.

## Greptile Summary

This PR fixes a critical timestamp synchronization issue across the validator system. The core problem was that different parts of the system were using different sources for timestamps:

1. The explorer was using `getBlockTime` which estimated timestamps based on slot periods
2. The ledger was using raw system time
3. The clock sysvar had its own timestamp

The fix ensures all components use the bank's clock timestamp as the single source of truth, specifically:

- Modified `magicblock-api/src/slot.rs` to use `bank.clock().unix_timestamp` instead of system time
- Updated `magicblock-rpc/src/json_rpc_request_processor.rs` to prioritize actual block timestamps over estimates
- Added comprehensive tests in `test_clocks_match.rs` to verify timestamp consistency

This change is particularly important for:
- Accurate transaction replay
- Consistent timestamp reporting between explorer and ledger
- Proper clock sysvar behavior

## Confidence score: 4.5/5

1. This PR is safe to merge as it fixes a critical timestamp consistency issue without introducing new risks
2. High confidence due to comprehensive test coverage and the straightforward nature of the fix - using a single source of truth
3. Files needing attention:
   - `magicblock-api/src/slot.rs`: Verify no edge cases in timestamp handling
   - `test_clocks_match.rs`: Ensure all critical timestamp scenarios are covered

<sub>3 files reviewed, 2 comments</sub>
<sub>[Edit PR Review Bot Settings](https://app.greptile.com/review/github) | [Greptile](https://greptile.com?utm_source=greptile_expert&utm_medium=github&utm_campaign=code_reviews&utm_content=magicblock-validator_454)</sub>

### CODE REVIEW

File: magicblock-api/tests/test_clocks_match.rs [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+                tx_result,
+                current_slot,
+                SanitizedTransaction::from_transaction_for_tests(transaction),
+                TransactionStatusMeta::default(),
+                current_slot as usize,
+            )
+            .unwrap();
+
+        // Wait for processing
+        tokio::time::sleep(Duration::from_millis(millis_per_slot * 110 / 100))
```
[the remark this answers is not in the document]
Comment (Dodecahedr0x) [reply magicblock-api/tests/test_clocks_match.rs:115]: I wait a little bit more than a slot to make sure the slot is processed

**Comment by Dodecahedr0x** [comment]
> Also, we should add a test to the ledger-restore integration tests like I said when we had a call. In there, you should make sure that if I run a transaction that relies on the clock to pass, it will pass again on ledger replay, making sure that the clock values match both in the original run and during the replay run.

Addressed in [7b3d039](https://github.com/magicblock-labs/magicblock-validator/pull/454/commits/7b3d03971ef9cd9f7b7c9a6673a2c66a4ac63056) (and fixed in [32722ac](https://github.com/magicblock-labs/magicblock-validator/pull/454/commits/32722acc7a69f6692a43c601b3a2bbfcbd078995) and [008f3e2](https://github.com/magicblock-labs/magicblock-validator/pull/454/commits/008f3e293c3c716e0808dfe35dd04752ccbec351)). I could rewrite the git history to make it a single commit for ease of review

File: magicblock-api/tests/test_clocks_match.rs [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+    system_instruction,
+    transaction::{SanitizedTransaction, Transaction},
+};
+use tempfile::TempDir;
+
+/// Test that verifies transaction timestamps, block timestamps, and ledger block timestamps all match
+#[tokio::test]
+async fn test_clocks_match() -> ApiResult<()> {
+    // Initialize a temporary directory for the test
+    let temp_dir = TempDir::new()?;
```
[the remark this answers is not in the document]
Comment (Dodecahedr0x) [reply magicblock-api/tests/test_clocks_match.rs]: Addressed in [fa26183](https://github.com/magicblock-labs/magicblock-validator/pull/454/commits/fa261839a8fa93e60c20a360269b5094dca09473)

File: test-integration/test-runner/bin/run_tests.rs [hunk 3]
```diff
@@ -391,14 +391,30 @@ fn run_magicblock_api_tests(
     if !should_run_test("magicblock_api") {
         return Ok(success_output());
     }
+    let loaded_chain_accounts =
```
[the remark this answers is not in the document]
Comment (Dodecahedr0x) [reply test-integration/test-runner/bin/run_tests.rs]: addressed in [c6d9987](https://github.com/magicblock-labs/magicblock-validator/pull/454/commits/c6d998740386fa60c79104809135f07ed1fae38d)

File: magicblock-api/src/slot.rs [hunk 4]
```diff
@@ -24,15 +22,8 @@ pub fn advance_slot_and_update_ledger(
     // Update ledger with previous block's metas
     let ledger_result = ledger.write_block(
         prev_slot,
-        timestamp_in_secs() as i64,
+        bank.clock().unix_timestamp,
```
[the remark this answers is not in the document]
Comment (Dodecahedr0x) [reply magicblock-api/src/slot.rs]: addressed in [bd3637f](https://github.com/magicblock-labs/magicblock-validator/pull/454/commits/bd3637fc8466c096c879a88aec13986f209435cf)
